params.go

```diff
@@ -0,0 +1,157 @@
+package slog
+
+import (
+	"context"
+	"sync"
+	"testing"
+)
+
+// WithParams returns a copy of the parent context containing the given log parameters.
+// Any log events generated using the returned context will include these parameters
+// as metadata.
+//
+// For example:
+//
+//	 ctx := slog.WithParams(ctx, map[string]string{
+//	   "foo_id": fooID,
+//	   "bar_id": barID,
+//	 })
+//
+//	slog.Info(ctx, "Linking foo to bar")  // includes foo_id and bar_id parameters
+//
+// If the parent context already contains parameters set by a previous call to WithParams,
+// the new parameters will be merged with the existing set, with newer values taking
+// precedence over older ones.
+//
+// It is not safe to modify the supplied map after passing it to WithParams.
+func WithParams(parent context.Context, params map[string]string) context.Context {
+	return context.WithValue(parent, contextKeyParamNode{}, &paramNode{
+		Parent:      parent,
+		ChildParams: params,
+	})
+}
+
+// WithParam is shorthand for calling WithParams with a single key-value pair.
+func WithParam(ctx context.Context, key, value string) context.Context {
+	return WithParams(ctx, params{key: value})
+}
+
+// Params returns all parameters stored in the given context using WithParams. This
+// function is intended to be used by libraries _other_ than slog that want access to the
+// set of parameters (i.e. `monzo/terrors` functions).
+//
+// The return value is guaranteed to be non-nil and can be safely mutated by the caller.
```

Comment:
> and can be safely mutated by the caller.

Guaranteeing this has a performance cost as called out below: even when the result of `Params` is cached in the `paramNode`, we must clone it before returning it to the user. On balance I still think it's better to bias towards safety here.